### DISCUSSION TITLE
Energy Estimation Accuracy with Input Tokens

### DISCUSSION THREAD

**Discussion by RameezI** [mlco2/ecologits#86] [Q&A] [answered]
Thank you for the great work on this project! I noticed that energy estimation currently focuses on output tokens and request latency. However, input tokens and context length significantly affect the compute load, especially in large language models (LLMs).

Is there a reason input tokens are not included in the energy estimation? Are there any plans to incorporate this aspect in future updates?

Appreciate your time and insights!

**Comment by adrienbanse** [comment, marked as answer]
Hey @RameezI, sorry for the late answer, and many thanks for your interest in this project!
Following how LLMs work, we make the assumption that the typical request involves much more decoding than encoding, and therefore that the energy consumption of the encoding is negligible. 
It is however true that we don’t handle the edge cases such as giving a huge text to a model, and ask for a very precise answer that would involve only a few tokens (for example, give a book to a model and ask for the age of a particular character). 
Could you send us the references that state that the input context length significantly affects the compute load? If it turns out that it has more impact than what we think, we'll try to update the methodology accordingly.

**Reply by thomasGuerin3** [reply]
Hi @adrienbanse ! 

> It is however true that we don’t handle the edge cases such as giving a huge text to a model, and ask for a very precise answer that would involve only a few tokens (for example, give a book to a model and ask for the age of a particular character).


I'm in this case (and it might be the case for every application that uses RAG systems) : 97% of my tokens are input tokens. 

So it seems odd to calculate the footprint of my llm application by relying solely on the output tokens.


> Could you send us the references that state that the input context length significantly affects the compute load?


The only study I could find on the subject is this one : https://arxiv.org/pdf/2407.04014

In the study, there are two figures that I find interesting : 

<img width="304" alt="image" src="https://github.com/user-attachments/assets/5da6be89-b361-40a4-a926-8e39da5a231f" />


As you can see there's an asymmetry in Energy Scaling : 

**Output**: Energy per token tends to increase with longer outputs.

**Input**: Energy per token tends to decrease with longer inputs, but the order of magnitude (compared to the output graph) makes me think it's not negligible though.



### What can we do ?

I'm wondering if a price based approach could be used to estimate the impact of the input tokens.

For example, with gpt-4o : 

> Input:
$2.50 / 1M tokens
Output:
$10.00 / 1M tokens


Let's say I have 5000 input tokens and 50 output tokens.

According to the huggingface calculator : 

50 output tokens = **4.39 Wh** = 0,0878 Wh/token

The OpenAI pricing suggests that input tokens are 4 times less consuming that output tokens. 

So an estimate for the input tokens could be : 5000 * 0,0878 / 4 = **109,75 Wh**


But, knowing that the energy per token tends to decrease with longer inputs, that doesn't seems right either.. 🤔

**Reply by adrienbanse** [reply]
Hi @thomasGuerin3 👋 

> I'm in this case (and it might be the case for every application that uses RAG systems) : 97% of my tokens are input tokens.
> 
> So it seems odd to calculate the footprint of my llm application by relying solely on the output tokens.

The edge case I was referring to is the one where **the number of pass in the network is comparable for the input and the output**, that is when the answer contains just a few output tokens. But even if you have a lot of input tokens (say 2000) and an answer of, let's say 100 output tokens, there is ~100x more passes in the network for the output than the input (hence much more energy), even though there is 20x more input tokens than output tokens. 

> The only study I could find on the subject is this one : https://arxiv.org/pdf/2407.04014
> 
> In the study, there are two figures that I find interesting :
> 
> <img alt="image" width="304" src="https://private-user-images.githubusercontent.com/43245220/430284197-5da6be89-b361-40a4-a926-8e39da5a231f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM4MzgyNjUsIm5iZiI6MTc0MzgzNzk2NSwicGF0aCI6Ii80MzI0NTIyMC80MzAyODQxOTctNWRhNmJlODktYjM2MS00MGE0LWE5MjYtOGUzOWRhNWEyMzFmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDA1VDA3MjYwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFjOTI2NTEwMGQyNjhiMzUwNzBjOTI1YmY3NGVlZmQ1N2M0YTQ1ZDMwNGIyMjBjNTZmOGY2NjlmZjA1MDQ1ODkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.hZc35IF41mUVEXmrQEMTK3VviZMSmtMkirR8UcnzjPA">
> As you can see there's an asymmetry in Energy Scaling :
> 
> **Output**: Energy per token tends to increase with longer outputs.
> 
> **Input**: Energy per token tends to decrease with longer inputs, but the order of magnitude (compared to the output graph) makes me think it's not negligible though.

I think that this plot is compatible with my understanding. The total energy doesn't change much with the number of input tokens, so it makes sense that the energy / token decreases with the number of tokens... I expect this to be a slightly increasing line (almost a constant) if you plot energy vs. input token and not energy / token vs. input token. 

Concerning the order of magnitude, the authors say that they maintain the # of output tokens to 32. The plotted energy therefore also accounts for the generation of these 32 output tokens...

> 
> ### What can we do ?
> I'm wondering if a price based approach could be used to estimate the impact of the input tokens.
> 
> For example, with gpt-4o :
> 
> > Input:
> > $2.50 / 1M tokens
> > Output:
> > $10.00 / 1M tokens

The only thing that makes me doubt is this pricing.. I cannot really understand it, because it seems indeed that the number of input tokens is non negligible. It could be that this pricing is such as it is for other reasons than energy consumption. It could also be that I'm wrong and that I'm missing something. 😅

@samuelrince Do you have any insight on this? If it turns out that it has a non negligible effect, then a method based on the pricing is, imo, interesting! Thanks for the idea @thomasGuerin3 . :)

**Comment by emerzon** [comment]
LLM Inference is split into 2 phases, PP (prompt processing) and TG (token generation).
Not considering input tokens discards completely the PP phase. 
There are several benchmarks for opensource models on the PP/TG performance.
Ie. https://github.com/ggml-org/llama.cpp/discussions/15013

The PP/TG ratio varies a lot according to model arch and hardware, but 30:1 seems to be a very rought acceptable average. So you could also infer that for power usage.
